kernel: Option to assert on spin lock time

teburd opened the pull request:
Spin locks held for any lengthy duration prevent interrupts and in a real time system where interrupts drive tasks this can be problematic. Add an option to assert if a spin lock is held for a duration longer than the configured number of cycles.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
kernel: Option to assert on spin lock time

Spin locks held for any lengthy duration prevent interrupts and
in a real time system where interrupts drive tasks this can be
problematic. Add an option to assert if a spin lock is held for
a duration longer than the configurable number of microseconds.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
A draft for now. Honestly would be nice to expose the asserts such that the call site is provided, that would require making k_spin_lock() k_spin_unlock() macros though I believe, or at least macro wrappers. I've done that several times now locally and I'd really rather prefer to have an option to enable that, along with some wait time and lock time checks to validate spin locks aren't waiting or being held for significant durations.

Hopeful some discussion comes from this initial stab at adding part of that.

teburd commented:
Pushing off the idea to switch lock/unlock to be macros to make debugging easier for a next step, add the spin lock time limit debugging option and assert here though.

andyross commented on include/zephyr/spinlock.h:
```diff
@@ -183,7 +192,16 @@ static ALWAYS_INLINE void k_spin_unlock(struct k_spinlock *l,
+	/* Verify the time limit has not been exceeded ignoring
+	 * the rule if there was a counter wrap.
+	 */
+	__ASSERT(k_us_to_cyc_floor32(CONFIG_SPIN_LOCK_TIME_LIMIT) >
```
That conversion is runtime code on some platforms (really just HPET I think, where the cycle rate is probed and unknown at build time), which turns this into a pretty heavyweight bit of intstrumentation.  Might be cleaner just to specify it in cycles.  Real time units are nice when you can have them, but anyone who cares about tuning at this level will know how to read a cycle time.

andyross commented on subsys/debug/Kconfig:
```diff
@@ -261,6 +261,14 @@ config SPIN_VALIDATE
 	  kernel code size, don't use on platforms known to be small.
 
+config SPIN_LOCK_TIME_LIMIT
+	int "Spin lock holding time limit in microseconds"
```
Microseconds seems too coarse to me.  Code that lives in this world tends to think a 10us delay is just barely tolerable.

teburd commented on include/zephyr/spinlock.h:
```diff
@@ -152,7 +158,10 @@ static ALWAYS_INLINE k_spinlock_key_t k_spin_lock(struct k_spinlock *l)
 	z_spin_lock_set_owner(l);
-#endif
+#if defined(CONFIG_SPIN_LOCK_TIME_LIMIT) && (CONFIG_SPIN_LOCK_TIME_LIMIT != 0)
+	l->lock_time = sys_clock_cycle_get_32();
```
It's really quite difficult to instrument the spin lock time without actually using spin locks in the timer drivers it seems, looking at even a select few timer drivers I'm wondering why a spin lock is even needed on getting the current cycle count. In most cases it should be a register read. Unfortunate but right, I might need to go add some code to all the timer drivers to get a lock-less read of the cycle count

teburd replied:
I've added a hidden Kconfig timer drivers may select to inform that the cycle count accessors are lock free, and can be used directly with this.

Commit:
timer: cAVS cycle count is lock free

Select the Kconfig option noting that the cavs (intel adsp)
timer driver provides a lock free cycle count accessor and
therefore can be used with a spin lock time limit assert.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
timer: APIC timer count is a lock free accessor

Selects the hidden Kconfig to note that the APIC timer's cycle count
accessor is a lock free path and may use the spin lock time limit check.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

andyross reviewed (approved):
One quibble about driver selection, but nothing that should block merge; we can always audit drivers later.

evgeniy-paltsev commented on include/zephyr/spinlock.h:
```diff
@@ -183,7 +192,15 @@ static ALWAYS_INLINE void k_spin_unlock(struct k_spinlock *l,
-#endif
+
+#if defined(CONFIG_SPIN_LOCK_TIME_LIMIT) && (CONFIG_SPIN_LOCK_TIME_LIMIT != 0)
+	uint32_t delta = sys_clock_cycle_get_32() - l->lock_time;
```
Hmm, is it OK that we don't care about rollover here? `***cycle_get_32` represents the fastest system timer and on platforms with multi GHz clock rollover happens often.

teburd replied:
If a spin lock is held for more than one roll over of the timer clock then I think you've probably held a spin lock for too long already. This would be, at 4GHz, about a full second.

evgeniy-paltsev replied:
Ok, just imagine the situation:

* we lock the spinlock. The sys_clock_cycle_get_32 returns `0xFFFF'FFF0`, so we write this value to `l->lock_time`
* we unlock spinlock after 50 cycles. The sys_clock_cycle_get_32 returns `0x0000'0022`, so `delta = 0x0000'0022 - 0xFFFF'FFF0`.

As I can see we'll get false positive here.

teburd replied:
delta will return 50 with the above math given unsigned integer math according to the C specs here. The only time this is an issue is if the count wraps multiple times, which won't be seen. But that's unlikely even with the fastest clock speeds today to happen in a time frame that spin locks should be held for.

teburd replied:
The standard actually says this is how unsigned math for C is supposed to work, so we're good! See 6.2.5 paragraph 9 for the C99 ISO spec

teburd commented:
@evgeniy-paltsev can you please re-review. Thanks!
